### ISSUE TITLE
[inductor] Wrong argmax after `adaptive_avg_pool1d` when the pool input is a non-contiguous (transposed) view — CPU and CUDA

### ISSUE THREAD

**Issue by BUAA-Wander** [pytorch/pytorch#193492]
### 🐛 Describe the bug

`torch.compile` with the Inductor backend returns a wrong `argmax` index when
`torch.adaptive_avg_pool1d` is fused with `torch.argmax` and the pool **input is a
non-contiguous view** (e.g. produced by `transpose`). The eager result is correct;
the compiled result points to an index whose value is not the maximum. The mismatch
is deterministic (fixed input, no randomness).

The fused kernel reads the strided input as if it were contiguous, so it computes
argmax over the wrong data layout and reports a wrong index.

### To Reproduce

```python
import torch

def fn(x):
    x = torch.adaptive_avg_pool1d(input=x, output_size=2)
    return torch.argmax(input=x)

torch.manual_seed(123)
x = torch.rand([4, 4, 3], dtype=torch.float64).transpose(0, 1)  # non-contiguous

eager = fn(x.clone())
compiled = torch.compile(fn, backend="inductor")(x.clone())

print("eager   =", eager.item())     # 15
print("compiled=", compiled.item())  # 27 (wrong: value at 27 is not the max)

# pooled max location:
p = torch.adaptive_avg_pool1d(input=x, output_size=2)
print("true max at", torch.argmax(p).item(), "value", p.flatten().max().item())
```

Result (torch 2.13.0+cu130, CPU and CUDA):

```
eager    = 15
compiled = 27
true max at 15 value 0.8210611035776498
```

Notes:
- `x.transpose(0, 1).contiguous()` makes the compiled result correct (15).
- Returning both `(pooled, argmax)` instead of only `argmax` also gives the correct
  result (the pooled tensor is materialized and the fused kernel is not used).
- Reproduces with `output_size=3` as well; also reproduces on torch 2.5.1.
- Not reproduced for `adaptive_avg_pool2d + argmax` in the shapes tested.

### Root cause (generated kernel)

The fused C++ kernel `cpp_fused__adaptive_avg_pool2d_argmax_squeeze_unsqueeze_0`
indexes the transposed input with contiguous strides:

```cpp
auto tmp10 = in_ptr0[3L*x0 + div_floor(3L*x1, 2)];
auto tmp17 = in_ptr0[1L + 3L*x0 + div_floor(3L*x1, 2)];
...
tmp_acc0 = argmax_combine(tmp_acc0, tmp27, 2L*x0 + x1);
out_ptr0[0] = tmp_acc0.index;
```

The transposed input has shape `(4, 4, 3)` but strides `(4, 12, 1)`; the kernel
reads with a `(3, 1)`-style contiguous layout, i.e. it processes the wrong logical
tensor and returns an index into that wrong interpretation. When the transposed dims
have equal sizes the reads stay in bounds, which hides the bug.

### Versions

- PyTorch: 2.13.0+cu130 and 2.5.1+cu121 (both reproduce; CPU and CUDA)
- OS: Linux (AutoDL container)

cc @chauhang @penguinwu

**Comment by BUAA-Wander** [comment]
> I reproduced this on torch 2.10.0+cpu with the exact reporter input: eager returns 15 while Inductor returns 27 for a transposed `(4, 4, 3)` tensor with strides `(3, 12, 1)`.
> 
> I traced the failure to the `_adaptive_avg_pool2d` Inductor lowering used by `adaptive_avg_pool1d`. The generated fused kernel assumes a contiguous interpretation of the leading dimensions when the pooled result is fused directly into `argmax`.
> 
> A minimal fix I verified locally is to require contiguous storage before constructing the adaptive pooling pointwise loader:
> 
>  @register_lowering(aten._adaptive_avg_pool2d)
>  def _adaptive_avg_pool2d(x, output_size):
>      ...
>      assert isinstance(x, TensorBox)
>      assert len(output_size) == 2
>      x.realize_hint()
> +    x = ir.ExternKernel.require_contiguous(x)
> With caches disabled, the reporter repro changes from:
> 
> ```
> eager 15
> compiled 27
> ```
> 
> to:
> 
> ```
> eager 15
> compiled 15
> ```
> 
> This is deliberately conservative: it fixes correctness for non-contiguous inputs but may introduce a copy where a stride-aware lowering could avoid one. The regression test should cover the transposed 1D adaptive-pool + argmax case and verify both result and non-contiguous input layout. If maintainers prefer a stride-preserving fix rather than the contiguous guard, I can narrow the faulty index flattening further.

Thanks for the quick diagnosis and the patch! I ran a sibling-entry probe
around this fused-kernel path (1D/2D/3D adaptive pooling, other index
consumers, other non-contiguous constructors) on torch 2.13.0 and 2.5.1,
CPU and CUDA, fixed seed, caches disabled. Two observations that may be
useful for the fix and the regression test.

**1. The `_adaptive_avg_pool2d` path also affects 2D pooling and several 1D
variants.**

All of the following mismatch deterministically (eager correct, compiled
index points at a non-maximum pooled value), with identical indices on
CPU/CUDA and 2.13.0/2.5.1:

```text
adaptive_avg_pool1d (4,4,3)   strides (3,12,1)    out=2  argmax  eager 15  compiled 27
adaptive_avg_pool1d (4,4,3)   strides (3,12,1)    out=2  argmin  eager 28  compiled 22
adaptive_avg_pool1d (4,4,3)   strides (3,12,1)    out=3  argmax  eager 26  compiled  8
adaptive_avg_pool1d (4,4,3)   strides (3,12,1)    f32    argmax  eager 12  compiled 18
adaptive_avg_pool1d (3,4,4)   strides (16,1,4)    out=2  argmax  eager 19  compiled 21   # non-contiguous last dim
adaptive_avg_pool2d (4,4,5,4) strides (20,80,4,1) out=(2,2) argmax eager 59  compiled 47
```

The 1D case goes through `unsqueeze -> _adaptive_avg_pool2d -> squeeze` and
the 2D `out=(2,2)` case is a direct `_adaptive_avg_pool2d`, so a contiguous
guard at that lowering should cover all of the above.

**2. The same stride-blind indexing also shows up in the generic
reduction + argmax fusion — which the `_adaptive_avg_pool2d` guard will not
fix.**

`adaptive_avg_pool2d/3d` with `output_size = 1` decompose in the frontend
into `mean.dim([-1,-2] / [-1,-2,-3], keepdim=True) + argmax`, so they never
reach `_adaptive_avg_pool2d`. Plain `torch.argmax(torch.mean(...))` on the
same non-contiguous inputs already mismatches, with no pooling involved:

```text
adaptive_avg_pool2d out=(1,1) + argmax      (4,4,5,4)   strides (20,80,4,1)   eager 14  compiled 11
adaptive_avg_pool3d out=(1,1,1) + argmax    (4,4,3,3,3) strides (27,108,9,3,1) eager  2  compiled  8
mean(dim=(-1,-2)) + argmax                  (4,4,5,4)   strides (20,80,4,1)   eager 14  compiled 11
mean(dim=-1) + argmax                       (4,4,3)     strides (3,12,1)      eager  7  compiled 13
sum(dim=(-1,-2)) + argmax                   (4,4,5,4)   strides (20,80,4,1)   eager 14  compiled 11
amax(dim=(-1,-2)) + argmax                  (4,4,5,4)   strides (20,80,4,1)   eager 12  compiled  3
```

The reduction values themselves match eager exactly when argmax is not fused
(unfused sanity check), so the bug is in the fused kernel's input indexing,
not in the reduction math. The generated kernel `cpp_fused_argmax_mean_0`
reads `in_ptr0[x1 + 20*x0]`, i.e. it flattens the leading dims and the
reduction window as if the input were contiguous, ignoring the real strides
`(20, 80, 4, 1)` — the same assumption you identified, but in the generic
reduction fusion rather than the adaptive-pool lowering. That seems
consistent with your note that the faulty index flattening could be narrowed
one level up.

Minimal repro without any pooling:

```python
import torch

def fn(x):
    return torch.argmax(torch.mean(x, dim=(-1, -2), keepdim=True))

torch.manual_seed(123)
x = torch.rand([4, 4, 5, 4], dtype=torch.float64).transpose(0, 1)  # strides (20, 80, 4, 1)

print("eager   =", fn(x.clone()).item())                 # 14
print("compiled=", torch.compile(fn)(x.clone()).item())  # 11 (value at 11 is not the max)
```

To me these look like the same "fused kernel assumes contiguous leading dims"
mechanism reached through different lowerings. If you agree, a guard or
regression test at the reduction-fusion level would cover both; otherwise I
can open a separate issue for the reduction path. Happy to provide any of the
probe scripts.

**Comment by sylvesterkaczmarek** [comment]
Opened #194093 with a focused Inductor fix and regression coverage. The adaptive average pooling input is constrained to contiguous layout before lowering, which covers both the pooled case and the identity-size early return from the follow-up repro. The separate generic non-contiguous reduction + `argmax` cases still reproduce independently, so I kept those outside this patch.

**Comment by sylvesterkaczmarek** [comment]
I reproduced this issue and traced it to Inductor's `_adaptive_avg_pool2d` path, which is also used by `adaptive_avg_pool1d`.

For non-contiguous inputs such as transposed views, the fused path can effectively interpret the input with the wrong layout when the pooled result is consumed directly by an arg reduction. This produces a deterministic incorrect index even though eager execution is correct.

Follow-up testing confirmed that the affected `_adaptive_avg_pool2d` path includes the original `adaptive_avg_pool1d(..., output_size=2) + argmax` case, `output_size=3`, `argmin`, alternate non-contiguous strides, and a direct 2D adaptive-pooling case.

I opened #194093 with a focused fix. It applies Inductor's existing contiguous layout constraint to `_adaptive_avg_pool2d` before the lowering executes, so the constraint also covers early-exit paths such as the identity-size case. The PR also adds regression coverage for the transposed `adaptive_avg_pool1d + argmax` cases with output sizes 2 and 3.

I verified the original CPU repro before the fix as eager `15` versus compiled `27`; with the fix, the tested adaptive-pooling cases match eager.

The separate generic non-contiguous reduction + `argmax` cases involving `mean`, `sum`, and `amax` reproduce independently and are outside the scope of #194093, so those can be handled separately if needed.

PR: #194093

**Comment by eellison** [comment]
i'm going to close as dupe of https://github.com/pytorch/pytorch/issues/193661 . slightly different, but we'll make sure the fix encompasses both
